### DISCUSSION TITLE
Load workflow instance by ID and get available actions for the active tasks

### DISCUSSION THREAD

**Discussion by mzand111** [elsa-workflows/elsa-core#2661] [Q&A] [answered]
Hi
I’m new to Elsa. First of all I should thank you for the provided solution. I think Elsa has not yet received the attention it deserves. Even it is not found easily when searching. Maybe It needs more contents or even rebranding under a new name. 
Anyway, I am looking for a way to load workflow instance and get available actions in the current state. To be more specific in the provided sample ‘Elsa.Samples.UserTaskConsole’ consider that the user needs to load a previously halted instance by the workflowInstanceId and then list the available actions if the current activity is a user task so that user can select and trigger the desired one.
Thanks in advance

**Comment by sfmskywalker** [comment, marked as answer]
Hi @mzand111 , thank you for the feedback!

The UserTask Activity project is a little underdeveloped as it stands right now (it is still lacking convenient APIs to easily query available user actions for a given workflow instance).

However, you can still do it using the `IBookmarkFinder` and `IBookmarkSerializer` lower-level APIs. For example, let's say you have a workflow instance ID and you now want to get the available user actions:

```csharp
var workflowInstanceId = "some workflow instance ID";

// Get the bookmark typename
var bookmarkTypeName = typeof(UserTaskBookmark).GetSimpleAssemblyQualifiedName();
var bookmarks= await _bookmarkFinder.FindBookmarksByTypeAsync(bookmarkTypeName);
var userTaskBookmarks = bookmarks.Select(bookmark => _bookmarkSerializer.Deserialize<UserTaskBookmark>(bookmark.Model)).ToList();

var availableActions = userTaskBookmarks.Select(x => x.Action).ToList();
```

The above should be encapsulated into a new service called `IUserTaskService` to make it easier.

**Reply by sfmskywalker** [reply]
Actually, the above code snippet doesn't take into account the workflow instance ID - this is an oversight of the BookmarkFinder which we should update. Until that's done, you'd have to drop into a lower API: the `IBookmarkStore`.

**Reply by sfmskywalker** [reply]
I just pushed [an update](https://github.com/elsa-workflows/elsa-core/commit/a12f18f7add1151c59a8aace3f478557d276158d) that adds a new service called `IUserTaskService` that you can use like this:

```csharp
var userActions = await _userTaskService.GetUserActionsAsync(workflowInstanceId);
```

**Reply by mzand111** [reply]
That would be very useful for me. Thanks
